viabel/optimization_diagnostics.py

```diff
@@ -0,0 +1,256 @@
+
+import autograd.numpy as np
+from autograd.extend import primitive
+
+def compute_R_hat(chains, warmup=500):
+    #first axis is relaisations, second is iters
+    # N_realisations X N_iters X Ndims
+    jitter = 1e-8
+    chains = chains[:, warmup:, :]
+    n_iters = chains.shape[1]
+    n_chains = chains.shape[0]
+    K = chains.shape[2]
+    if n_iters%2 == 1:
+        n_iters = int(n_iters - 1)
+        chains = chains[:,:n_iters-1,:]
+
+    n_iters = n_iters // 2
+    psi = np.reshape(chains, (n_chains * 2, n_iters, K))
+    n_chains2 = n_chains*2
+    psi_dot_j = np.mean(psi, axis=1)
+    psi_dot_dot = np.mean(psi_dot_j, axis=0)
+    s_j_2 = np.sum((psi - np.expand_dims(psi_dot_j, axis=1)) ** 2, axis=1) / (n_iters - 1)
+    B = n_iters * np.sum((psi_dot_j - psi_dot_dot) ** 2, axis=0) / (n_chains2 - 1)
+    W = np.nanmean(s_j_2, axis=0)
+    W = W + jitter
+    var_hat = (n_iters - 1) / n_iters + (B / (n_iters*W))
+    R_hat = np.sqrt(var_hat)
+    return var_hat, R_hat
+
+def compute_R_hat_halfway(chains, interval=100, start=1000):
+    n_chains, n_iters, K= chains.shape
+    n_subchains = n_iters //interval
+    r_hats_halfway = list()
+
+    for i in range(n_subchains):
+        sub_chains = chains[:, :start+(i+1)*interval,:]
+        n_sub_chains, n_sub_iters, K = sub_chains.shape
+        r_hat_current = compute_R_hat(sub_chains, warmup=n_sub_iters//2)[1]
+        r_hats_halfway.append(r_hat_current)
+
+    return np.array(r_hats_halfway)
+
+def compute_R_hat_halfway_light(chains, warmup_fraction=0.5):
+    n_chains, n_iters, K= chains.shape
+    warmup = int(warmup_fraction*n_iters)
+    r_hat_current = compute_R_hat(chains, warmup=n_iters//2)[1]
+    return r_hat_current
+
+def monte_carlo_se(iterate_chains, warmup=500):
+    chains = iterate_chains[:, warmup:, :]
+    n_chains, N_iters, K = chains.shape[0], chains.shape[1], chains.shape[2]
+    mcse_combined_list = np.zeros((N_iters,K))
+    print(chains.shape)
+    for i in range(1,N_iters):
+        chains_sub = chains[:, :i,:]
+        n_chains, n_iters, K = chains_sub.shape[0], chains_sub.shape[1], chains_sub.shape[2]
+        chains_flat = np.reshape(chains_sub, (n_chains*i, K))
+        variances = np.var(chains, ddof=1, axis=1)
+        variances_combined = np.var(chains_flat, ddof=1, axis=0)
+        mean_var_chains = np.mean(variances, axis=0)
+        mcse_per_chain = np.sqrt(variances / n_iters)
+        mcse_combined = np.sqrt(variances_combined/(i*n_chains))
+        mcse_combined_list[i] = mcse_combined
+
+    #print(mcse_per_chain.shape)
+    return mcse_per_chain, mcse_combined_list
+
+# compute mcmcse for a chain/array
+def monte_carlo_se2(iterate_chains, warmup=500, param_idx=0):
```

Comment:
How is this different from monte_carlo_se2? Please use more descriptive names to distinguish these